File: .tmp_py4013/batch_04_sources/2512_reward_top_k_students.js
// LeetCode 2512 - Reward Top K Students
// https://leetcode.com/problems/reward-top-k-students/

/**
 * @param {string[]} positive_feedback
 * @param {string[]} negative_feedback
 * @param {string[]} report
 * @param {number[]} student_id
 * @param {number} k
 * @return {number[]}
 */
var topStudents = function(positive_feedback, negative_feedback, report, student_id, k) {
    const pos = new Set(positive_feedback);
    const neg = new Set(negative_feedback);
    const arr = Array(report.length);
    for (let i = 0; i < report.length; i++) {
        let score = 0;
        for (const w of report[i].split(' ')) {
            if (!w) continue;
            if (pos.has(w)) score += 3;
            else if (neg.has(w)) score--;
        }
        arr[i] = [student_id[i], score];
    }
    arr.sort((a, b) => a[1] !== b[1] ? b[1] - a[1] : a[0] - b[0]);
    const ans = Array(k);
    for (let i = 0; i < k; i++) ans[i] = arr[i][0];
    return ans;
};

File: .tmp_rs4013/rust_check_batch08/2512_reward_top_k_students.rs
struct Solution;
// LeetCode 2512 - Reward Top K Students
// https://leetcode.com/problems/reward-top-k-students/

use std::collections::HashSet;

impl Solution {
    pub fn top_students(
        positive_feedback: Vec<String>,
        negative_feedback: Vec<String>,
        report: Vec<String>,
        student_id: Vec<i32>,
        k: i32,
    ) -> Vec<i32> {
        let pos: HashSet<&str> = positive_feedback.iter().map(|s| s.as_str()).collect();
        let neg: HashSet<&str> = negative_feedback.iter().map(|s| s.as_str()).collect();
        let mut arr: Vec<(i32, i32)> = Vec::new();
        for i in 0..report.len() {
            let mut score = 0;
            for w in report[i].split_whitespace() {
                if pos.contains(w) {
                    score += 3;
                } else if neg.contains(w) {
                    score -= 1;
                }
            }
            arr.push((student_id[i], score));
        }
        arr.sort_by(|a, b| {
            if a.1 != b.1 {
                b.1.cmp(&a.1)
            } else {
                a.0.cmp(&b.0)
            }
        });
        arr.into_iter().take(k as usize).map(|(id, _)| id).collect()
    }
}

fn main() {}
